File: app/src/main/java/com/msaggik/algorithmsdatastructures/java/sort/HeapSort.java
package com.msaggik.algorithmsdatastructures.java.sort;

/**
 * Пирамидальная сортировка
 * <p>
 * Вычислительная сложность: О(N*log(N))
 * Затраты памяти: O(1)
 **/

public class HeapSort {
    public static int[] heapSort(int[] arrayInt) {
        // heap creation
        for(int i = arrayInt.length / 2 - 1; i >= 0; i--) {
            createHeap(arrayInt, i, arrayInt.length);
        }
        // building an array from the heap
        for(int i = arrayInt.length - 1; i >= 0; i--) {
            int value = arrayInt[0];
            arrayInt[0] = arrayInt[i];
            arrayInt[i] = value;
            createHeap(arrayInt, 0, i);
        }
        return arrayInt;
    }

    private static void createHeap(int[] array, int root, int length) {
        int left = 2 * root + 1;
        int right = 2 * root + 2;
        int middle = root;

        if (left < length && array[left] > array[middle]) middle = left;
        if (right < length && array[right] > array[middle]) middle = right;

        if (middle != root)  {
            int value = array[middle];
            array[middle] = array[root];
            array[root] = value;

            createHeap(array, middle, length);
        }
    }
}

File: app/src/main/java/com/msaggik/algorithmsdatastructures/kotlin/sort/HeapSort.kt
package com.msaggik.algorithms.kotlin.sort

object HeapSort {
    fun heapSort(arrayInt: IntArray): IntArray {
        // heap creation
        for (i in arrayInt.size / 2 - 1 downTo 0) {
            createHeap(arrayInt, i, arrayInt.size)
        }
        // building an array from the heap
        for (i in arrayInt.indices.reversed()) {
            val value = arrayInt[0]
            arrayInt[0] = arrayInt[i]
            arrayInt[i] = value
            createHeap(arrayInt, 0, i)
        }
        return arrayInt
    }

    private fun createHeap(array: IntArray, root: Int, length: Int) {
        val left = 2 * root + 1
        val right = 2 * root + 2
        var middle = root

        if (left < length && array[left] > array[middle]) middle = left
        if (right < length && array[right] > array[middle]) middle = right

        if (middle != root) {
            val value = array[middle]
            array[middle] = array[root]
            array[root] = value

            createHeap(array, middle, length)
        }
    }
}